file_zbc.c

```diff
@@ -0,0 +1,2240 @@
+			} else if (*zone_type != zone->type) {
+				*zone_type = 0;
+				return -EIO;
+			}
+		}
+
+		if (zone->type != BLK_ZONE_TYPE_CONVENTIONAL &&
+		    lba >= zone->wp) {
+			/* Read zeros */
+			bytes = lba_count * zdev->lba_size;
+			memset(buf, 0, bytes);
+			break;
+		}
+
+		if (zone->type != BLK_ZONE_TYPE_CONVENTIONAL)
+			boundary = zone->wp;
+		else
+			boundary = zone->start + zone->len;
+		if (lba + nr_lbas > boundary)
+			count = boundary - lba;
+		else
+			count = lba_count;
+		bytes = count * zdev->lba_size;
+
+		/* Read written data */
+		ret = pread(zdev->fd, buf, bytes,
+			    zdev->meta_size + lba * zdev->lba_size);
+		if (ret != bytes) {
+			tcmu_dev_err(zdev->dev, "Read failed %zd / %zu B\n",
+				     ret, bytes);
+			return -EIO;
+		}
+
+		lba_count -= count;
+		lba += count;
+		buf += bytes;
+
+	}
+
+	return nr_lbas;
+}
+
+/*
+ * Read command emulation.
+ */
+static int zbc_read(struct tcmu_device *dev, struct tcmulib_cmd *cmd)
+{
+	struct zbc_dev *zdev = tcmu_get_dev_private(dev);
+	uint8_t *cdb = cmd->cdb;
+	uint64_t lba = tcmu_get_lba(cdb);
+	struct iovec *iovec = cmd->iovec;
+	size_t iov_cnt = cmd->iov_cnt;
+	int zone_type = 0;
+	ssize_t ret;
+	int i;
+
+	tcmu_dev_dbg(dev, "Read LBA %llu+%u, %zu vectors\n",
+		     (unsigned long long)lba,
+		     tcmu_get_xfer_length(cdb), iov_cnt);
+
+	/* Check LBA and length */
+	ret = zbc_check_rdwr(dev, cmd);
+	if (ret != SAM_STAT_GOOD)
+		return ret;
+
+	for (i = 0; i < iov_cnt; i++) {
+		ret = __zbc_read(zdev, iovec[i].iov_base,
+				 iovec[i].iov_len / zdev->lba_size, lba,
+				 &zone_type);
+		if (ret <= 0) {
+			if (!zone_type)
+				return tcmu_set_sense_data(cmd->sense_buf,
+						   ILLEGAL_REQUEST,
+						   ASC_ATTEMPT_TO_READ_INVALID_DATA,
+						   NULL);
+			return tcmu_set_sense_data(cmd->sense_buf,
+						   MEDIUM_ERROR,
+						   ASC_READ_ERROR,
+						   NULL);
+		}
+		lba += ret;
+	}
+
+	return SAM_STAT_GOOD;
+}
+
+/*
+ * Write command emulation.
+ */
+static int zbc_write(struct tcmu_device *dev, struct tcmulib_cmd *cmd)
+{
+	struct zbc_dev *zdev = tcmu_get_dev_private(dev);
+	uint8_t *cdb = cmd->cdb;
+	uint64_t lba = tcmu_get_lba(cdb);
+	size_t nr_lbas = tcmu_get_xfer_length(cdb);
+	size_t iov_cnt = cmd->iov_cnt;
+	struct iovec *iovec = cmd->iovec;
+	size_t remaining = tcmu_iovec_length(iovec, iov_cnt);
+	struct blk_zone *zone;
+	off_t offset;
+	ssize_t ret;
+
+	tcmu_dev_dbg(dev, "Write LBA %llu+%u, %zu vectors\n",
+		     (unsigned long long)lba,
+		     tcmu_get_xfer_length(cdb), iov_cnt);
+
+	/* Check LBA and length */
+	ret = zbc_check_rdwr(dev, cmd);
+	if (ret != SAM_STAT_GOOD)
+		return ret;
+
+	/* For writes, check zone boundary crossing */
+	zone = zbc_get_zone(zdev, lba, false);
+	if (lba + nr_lbas > zone->start + zone->len) {
+		tcmu_dev_err(dev, "Write boundary violation lba %llu, xfer len %lu\n",
+			     lba, nr_lbas);
+		return tcmu_set_sense_data(cmd->sense_buf,
+					   ILLEGAL_REQUEST,
+					   ASC_WRITE_BOUNDARY_VIOLATION,
```

Comment:
Don't forget to include this addition to scsi_def.h.

Reply:
Ignore this comment. I was going to ask you to move the asc/ascq defs to scsi_def.h, but then flip flopped and was ok just leaving them in this file for now since you will be the only user.